sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java

```diff
@@ -485,8 +486,8 @@ public long getSplitBacklogBytes() {
    * The checkpoint for an unbounded {@link CountingSource} is simply the last value produced. The
    * associated source object encapsulates the information needed to produce the next value.
    */
-  @DefaultCoder(AvroCoder.class)
-  public static class CounterMark implements UnboundedSource.CheckpointMark {
+  @DefaultCoder(SerializableCoder.class)
```

Comment:
I think it should not be a breaking change

Reply:
Right but I don't think Beam supports a hot upgrade. 
I don't see a way how to avoid such change.

Reply:
We should make a note about this (as well as the larger change) in CHANGES.md. 

Serialization isn't guaranteed to be stable either--if we're switching coders should we switch to a custom (or Beam-schema-based) coder here?

Reply:
It worries me a bit that there is no workaround. I suppose we could advise copying the old CountingSource into your own project and using that directly.

Reply:
> Serialization isn't guaranteed to be stable either--if we're switching coders should we switch to a custom (or Beam-schema-based) coder here?

Yes, we can switch to custom coder, but what will be the advantage over `SerializableCoder` in this case? I'm not sure that I understand right this "Serialization isn't guaranteed to be stable either".

Reply:
The serialization might not be stable when upgrading a pipeline between Java versions (or, technically, if this class ever gets changed or moved).

Reply:
Ahh, right. I guess a use case for that  - it should be a streaming pipeline that was restarted with different (upgraded) version of Beam. Do we commit (Beam) to support this?

Reply:
I don't know if it's a solid commitment (e.g. this is itself a breaking change), but we really try hard where feasible.

Reply:
I would recommend creating a custom coder which is binary compatible to the Avro based one. This is actually rather trivial, Avro does zigzag varint encoding and the bytes of both longs in CounterMark are just written after another without anything added.

Zigzag encoding is also trivial to implement, or using `org.apache.beam.vendor.grpc.v1p54p0.com.google.protobuf.CodedOutputStream.encodeZigZag64` (and Protobuf will always remain a core dependency):

```
  public static long encodeZigZag64(final long n) {
    return (n << 1) ^ (n >> 63);
  }
  public static long decodeZigZag64(final long n) {
    return (n >>> 1) ^ -(n & 1);
  }
```

And varint encoding / decoding is already implemented in `org.apache.beam.sdk.util.VarInt`.

Reply:
@mosche I think you have a point but it's overkill in this case. I'd stick with `SchemaCoder` as Robert suggested.